iai-callgrind-runner/src/runner/tool/logfile_parser.rs

```diff
@@ -79,51 +78,86 @@ pub trait LogfileParserT {
-    let old = old.into_iter().map(Some).chain(iter::repeat_with(|| None));
-    let mut res = vec![];
-    for (old, new) in iter::zip(old, new) {
-        let (costs_summary, old_pid, old_parent_pid) = if let (
-            Some(
-                old @ LogfileSummary {
-                    costs: Some(old_costs),
-                    ..
-                },
-            ),
-            Some(costs),
-        ) = (&old, &new.costs)
-        {
-            let costs = Some(CostsSummary::new(costs, Some(old_costs)));
-            (costs, Some(old.pid), old.parent_pid)
-        } else {
-            let costs = new
-                .costs
-                .as_ref()
-                .map(|costs| CostsSummary::new(costs, None));
-            (costs, None, None)
-        };
-        res.push(ToolRunSummary {
-            command: new.command.to_string_lossy().to_string(),
+impl LogfileSummary {
+    fn raw_into_tool_run(self) -> ToolRunSummary {
+        ToolRunSummary {
+            command: self.command.to_string_lossy().to_string(),
+            old_pid: None,
+            old_parent_pid: None,
+            pid: None,
+            parent_pid: None,
+            summary: self.fields.into_iter().collect(),
+            details: (!self.details.is_empty()).then(|| self.details.join("\n")),
+            error_summary: self.error_summary,
+            costs_summary: None,
+            log_path: self.log_path,
+        }
+    }
+    pub fn old_into_tool_run(self) -> ToolRunSummary {
+        let costs_summary = self
+            .costs
+            .as_ref()
+            .map(|x| CostsSummary::new(&Costs::empty(), Some(x)));
+        let old_pid = Some(self.pid);
+        let old_parent_pid = self.parent_pid;
+        ToolRunSummary {
             old_pid,
             old_parent_pid,
-            pid: new.pid,
-            parent_pid: new.parent_pid,
-            summary: new.fields.into_iter().collect(),
-            details: (!new.details.is_empty()).then(|| new.details.join("\n")),
-            error_summary: new.error_summary,
             costs_summary,
-            log_path: new.log_path,
+            ..self.raw_into_tool_run()
+        }
+    }
+
+    pub fn new_into_tool_run(self) -> ToolRunSummary {
+        let costs_summary = self.costs.as_ref().map(|x| CostsSummary::new(x, None));
+        let pid = Some(self.pid);
+        let parent_pid = self.parent_pid;
+        ToolRunSummary {
+            pid,
+            parent_pid,
+            costs_summary,
+            ..self.raw_into_tool_run()
+        }
+    }
+
+    pub fn merge(self, old: &LogfileSummary) -> ToolRunSummary {
+        assert_eq!(self.command, old.command);
+        let costs_summary = (self.costs.is_some() || old.costs.is_some()).then(|| {
+            let emp = Costs::empty();
+            CostsSummary::new(self.costs.as_ref().unwrap_or(&emp), old.costs.as_ref())
         });
+        let old_pid = Some(self.pid);
+        let old_parent_pid = self.parent_pid;
```

Comment:
This should be `let old_parent_pid = old.parent_pid`. Same issue with `old_pid` above.

Reply:
Sorry I missed this